lib/Dialect/TTNN/Transforms/TTNNLayout.cpp

```diff
@@ -0,0 +1,389 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TT/Utils/OperandConstraints.h"
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "ttmlir/Dialect/TTNN/Utils/Utils.h"
+
+namespace mlir::tt::ttnn {
+#define GEN_PASS_DEF_TTNNLAYOUT
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h.inc"
+
+// Default collapse dims for affine map (d0, d1, d2) -> (d0 <> d1, d2)
+static const std::array<std::pair<int64_t, int64_t>, 1> g_defaultCollapseDims =
+    {{{0, -1}}};
+
+// Default memory space for tensors on host
+static const BufferType g_defaultMemorySpaceHost = BufferType::SystemMemory;
+
+// Default memory space for tesnors on device
+static const BufferType g_defaultMemorySpaceDevice = BufferType::DRAM;
+
+// Default memory layout for tensors on device
+static const TensorMemoryLayout g_defaultMemoryLayout =
+    TensorMemoryLayout::Interleaved;
+
+//===----------------------------------------------------------------------===//
+// Helper methods
+//===----------------------------------------------------------------------===//
+
+inline Location appendInputSuffix(Location loc, int64_t operandIndex) {
+  if (isa<NameLoc>(loc)) {
+    NameLoc oldLoc = mlir::cast<NameLoc>(loc);
+    StringAttr newName = StringAttr::get(
+        loc->getContext(), oldLoc.getName().str() + "_in_" +
+                               std::to_string(operandIndex) + "_layout");
+
+    return NameLoc::get(newName, oldLoc.getChildLoc());
+  }
+
+  return loc;
+}
+
+//===----------------------------------------------------------------------===//
+// To layout pass
+//===----------------------------------------------------------------------===//
+
+// Converts tensor types to have a tensor config attribute with default values
+//
+// Example: tensor<15x10x32xf32> -> tensor<15x10x32xf32, tensor_config<...>>
+// where tensor_config<...> is constructed with default values
+// SystemMemory, MemoryLayout::None, Grid<1x1>
+class TTNNLayoutTensorTypeConverter : public TypeConverter {
+public:
+  TTNNLayoutTensorTypeConverter(MLIRContext *ctx, GridAttr deviceGrid) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx, deviceGrid](RankedTensorType type) -> Type {
+      auto layout = type.getEncoding();
+      if (layout) {
+        return type;
+      }
+
+      std::int64_t deviceGridRank = deviceGrid.getShape().size();
+
+      // Default to single core grid
+      auto tensorGrid = GridAttr::get(ctx, deviceGridRank);
+
+      llvm::ArrayRef<std::pair<int64_t, int64_t>> collapseDimsRef(
+          g_defaultCollapseDims);
+
+      auto newLayout = TensorConfigAttr::get(
```

Comment:
I would love to see proper types here instead of auto.

Reply:
Agree with you 100%. I see it throughout the code thats why I added it,  but in general I'm all for not using it...
I will fix it throught the file.

Reply:
I personally like `auto`, esp when the type is stated on the rhs :)

Reply:
@nsmithtt I meant generally in this file, take a look few lines above.
![image](https://github.com/user-attachments/assets/dc415136-a779-4d7a-b572-f6db79158c0b)
But even having it RHS its nice to have it still. :)